gcloud/datastore/api.py

```diff
@@ -147,17 +142,26 @@ def put(entities, connection=None):
 
     :type connection: :class:`gcloud.datastore.connection.Connection`
     :param connection: Optional connection used to connect to datastore.
+                       If not passed, inferred from the environment.
+
+    :type dataset_id: :class:`gcloud.datastore.connection.Connection`
+    :param dataset_id: Optional. The dataset ID used to connect to datastore.
+                       If not passed, inferred from the environment.
+
+    :raises: EnvironmentError if ``connection`` or ``dataset_id`` not passed,
+             and cannot be inferred from the environment.  ValueError if
+             one or more entities has a key with a dataset ID not matching
+             the passed / inferred dataset ID.
     """
     if not entities:
         return
 
-    connection = connection or _implicit_environ.CONNECTION
+    connection = _require_connection(connection)
+    dataset_id = _require_dataset_id(dataset_id)
```

Comment:
The thing we want to avoid is where all the entities are from `foo2` (valid data), there is no current `Batch` and `_implicit_environ.DATASET_ID == 'foo1'`.

------

The difference being that you could just take the first key or first entity and get **some** dataset ID without having to throw an `EnvironmentError`.

I'm unclear which is better:

```python
try:
    dataset_id = _require_dataset_id(dataset_id)
except EnvironmentError:
    dataset_id = entities[0].key.dataset_id
```

or

```python
dataset_id = _require_dataset_id(dataset_id or entities[0].key.dataset_id)
```

----

It seems like neither really does the trick fully, bu. I feel like the precedence should beThe difference being that you could just take the first key or first entity and get **some** dataset ID without having to throw an `EnvironmentError`.

I'm unclear which is better.

```python
try:
    dataset_id = _require_dataset_id(dataset_id)
except EnvironmentError:
    dataset_id = entities[0].key.dataset_id
```

or

```python
dataset_id = _require_dataset_id(dataset_id or entities[0].key.dataset_id)
```

It seems like neither really does the trick fully, but the `not in_batch` check later makes the first work "good enough". In terms of pulling an ID, the precedence should be

- `dataset_id` if passed explicitly
- `Batch.top.dataset_id`
- `entities[0].key.dataset_id`
- `_implicit_environ.DATASET_ID`